src/software/ai/hl/stp/tactic/goalie/goalie_fsm.cpp

```diff
@@ -77,82 +82,25 @@ Rectangle GoalieFSM::getNoChipRectangle(const Field &field)
-    // (width, pos_side, neg_side) and the line from the desired position to the
-    // center of the friendly goal
-    auto width_x_goal =
-        intersection(Line(goalie_desired_position, field.friendlyGoalCenter()),
-                     Line(goalie_restricted_area.posXPosYCorner(),
-                          goalie_restricted_area.posXNegYCorner()));
-    auto pos_side_x_goal =
-        intersection(Line(goalie_desired_position, field.friendlyGoalCenter()),
-                     Line(goalie_restricted_area.posXPosYCorner(),
-                          goalie_restricted_area.negXPosYCorner()));
-    auto neg_side_x_goal =
-        intersection(Line(goalie_desired_position, field.friendlyGoalCenter()),
-                     Line(goalie_restricted_area.posXNegYCorner(),
-                          goalie_restricted_area.negXNegYCorner()));
-
-    // if the goalie restricted area already contains the point, then we are
-    // safe to move there.
-    if (contains(goalie_restricted_area, goalie_desired_position))
-    {
-        return std::make_optional<Point>(goalie_desired_position);
-    }
-    // Due to the nature of the line intersection, its important to make sure the
-    // corners are included, if the goalies desired position intersects with width
-    // (see above), use those positions The last comparison is for the edge case when
-    // the ball is behind the net
-    else if (width_x_goal &&
-             width_x_goal->y() <= goalie_restricted_area.posXPosYCorner().y() &&
-             width_x_goal->y() >= goalie_restricted_area.posXNegYCorner().y() &&
-             field.friendlyGoalCenter().x() <= goalie_desired_position.x())
-    {
-        return std::make_optional<Point>(*width_x_goal);
-    }
+    // Default chip target is the enemy goal
+    Point chip_target = world.field().enemyGoalCenter();
 
-    // if either two sides of the goal are intercepted, then use those positions
-    else if (pos_side_x_goal &&
-             pos_side_x_goal->x() <= goalie_restricted_area.posXPosYCorner().x() &&
-             pos_side_x_goal->x() >= goalie_restricted_area.negXPosYCorner().x())
-    {
-        return std::make_optional<Point>(*pos_side_x_goal);
-    }
-    else if (neg_side_x_goal &&
-             neg_side_x_goal->x() <= goalie_restricted_area.posXNegYCorner().x() &&
-             neg_side_x_goal->x() >= goalie_restricted_area.negXNegYCorner().x())
+    std::vector<Circle> open_areas = findGoodChipTargets(world);
+    if (!open_areas.empty())
     {
-        return std::make_optional<Point>(*neg_side_x_goal);
+        chip_target = open_areas[0].origin();
+
+        // Avoid chipping out of field or towards friendly corners by clamping the chip
+        // target to the region in front of the friendly defense area
+        double clamped_chip_target_x =
+            std::clamp(chip_target.x(), world.field().friendlyDefenseArea().xMax(),
+                       world.field().enemyGoalCenter().x());
```

Comment:
why not iterate through the open_areas vector and see if we can find a better chip target instead of clamping it?

Reply:
didn't think of that, I'll try that out

Reply:
Decided to just pass in the target area to findGoodChipTargets and restrain our search for chip targets to the target area